scrapy/core/scheduler.py

```diff
@@ -34,6 +46,15 @@ def from_crawler(cls, crawler):
         return cls(dupefilter, jobdir=job_dir(settings), logunser=logunser,
                    stats=crawler.stats, pqclass=pqclass, dqclass=dqclass, mqclass=mqclass)
 
+    def request_key(self, request):
+        if 'scheduler_slot' in request.meta:
+            return request.meta['scheduler_slot']
+
+        return self._request_key(request)
+
+    def _request_key(self, request):
+        raise NotImplementedError
```

Comment:
If this method is required, it is better to use abstract classes (see https://docs.python.org/3/library/abc.html). 

We use twisted interfaces in a few other places, which solve a similar issue, but I'd prefer to move to abc everywhere (though I don't know what others think :)

Reply:
The idea was to mark this method as abstractmethod, so that it fails earlier

Reply:
Fixed.